client/deck-validator.js

```diff
@@ -67,174 +36,208 @@ function isCardInReleasedPack(packs, card) {
-    let requiredPlots = 7;
-    let isRains = false;
-    let extendedStatus = [];
-    let requiredDraw = 60;
-    let isValid = true;
-
-    if(_.any(deck.drawCards, card => {
-        return !card.card.faction_code;
-    })) {
-        status = 'Invalid';
-        extendedStatus.push('Deck contains invalid cards');
-
-        return { status: status, plotCount: plotCount, drawCount: drawCount, extendedStatus: extendedStatus, isValid: false };
-    }
+function rulesForBanner(faction, factionName) {
+    return {
+        mayInclude: card => card.faction_code === faction && !card.is_loyal && card.type_code !== 'plot',
+        rules: [
+            {
+                message: 'Must contain 12 or more ' + factionName + ' cards',
+                condition: deck => getDeckCount(_(deck.drawCards).filter(cardQuantity => cardQuantity.card.faction_code === faction)) >= 12
+            }
+        ]
+    };
+}
 
-    let combined = _.union(deck.plotCards, deck.drawCards);
+/**
+ * Validation rule structure is as follows. All fields are optional.
+ *
+ * requiredDraw  - the minimum amount of cards required for the draw deck.
+ * requiredPlots - the exact number of cards required for the plot deck.
+ * mayInclude    - function that takes a card and returns true if it is allowed in the overall deck.
+ * cannotInclude - function that takes a card and return true if it is not allowed in the overall deck.
+ * rules         - an array of objects containing a `condition` function that takes a deck and return true if the deck is valid for that rule, and a `message` used for errors when invalid.
+ */
+const agendaRules = {
+    // Banner of the stag
+    '01198': rulesForBanner('baratheon', 'Baratheon'),
+    // Banner of the kraken
+    '01199': rulesForBanner('greyjoy', 'Greyjoy'),
+    // Banner of the lion
+    '01200': rulesForBanner('lannister', 'Lannister'),
+    // Banner of the sun
+    '01201': rulesForBanner('martell', 'Martell'),
+    // Banner of the watch
+    '01202': rulesForBanner('thenightswatch', 'Night\'s Watch'),
+    // Banner of the wolf
+    '01203': rulesForBanner('stark', 'Stark'),
+    // Banner of the dragon
+    '01204': rulesForBanner('targaryen', 'Targaryen'),
+    // Banner of the rose
+    '01205': rulesForBanner('tyrell', 'Tyrell'),
+    // Fealty
+    '01027': {
+        rules: [
+            {
+                message: 'You cannot include more than 15 neutral cards in a deck with Fealty',
+                condition: deck => getDeckCount(_.filter(deck.drawCards, cardQuantity => cardQuantity.card.faction_code === 'neutral')) <= 15
+            }
+        ]
+    },
+    // Kings of Summer
+    '04037': {
+        cannotInclude: card => card.type_code === 'plot' && hasTrait(card, 'Winter'),
+        rules: [
+            {
+                message: 'Kings of Summer cannot include Winter plot cards',
+                condition: deck => !_.any(deck.plotCards, cardQuantity => hasTrait(cardQuantity.card, 'Winter'))
```

Comment:
Rules like these weren't strictly necessary, as the deck would be ruled invalid from the `cannotInclude` function. However, it's nice to have the reason the deck is invalid highly visible. Either way, the specific plot will be labelled as not allowed.